instrument/extended.go

```diff
@@ -21,26 +21,99 @@
 package instrument
 
 import (
+	"fmt"
 	"os"
 	"runtime"
+	"strings"
 	"sync/atomic"
 	"time"
 
-	"github.com/shirou/gopsutil/process"
+	"github.com/m3db/m3x/process"
+
 	"github.com/uber-go/tally"
 )
 
+// ExtendedMetricsType is a type of extended metrics to report.
+type ExtendedMetricsType int
+
+const (
+	// NoExtendedMetrics describes no extended metrics.
+	NoExtendedMetrics ExtendedMetricsType = iota
+
+	// SimpleExtendedMetrics describes just a simple level of extended metrics:
+	// - number of active goroutines
+	// - number of configured gomaxprocs
+	// - number of file descriptors
+	SimpleExtendedMetrics
+
+	// DetailedExtendedMetrics describes a detailed level of extended metrics:
+	// - number of active goroutines
+	// - number of configured gomaxprocs
+	// - number of file descriptors
+	// - memory allocated running count
+	// - memory used by heap
+	// - memory used by heap that is idle
+	// - memory used by heap that is in use
+	// - memory used by stack
+	// - number of garbage collections
+	// - GC pause times
+	DetailedExtendedMetrics
+
+	// DefaultExtendedMetricsType is the default extended metrics level.
+	DefaultExtendedMetricsType = DetailedExtendedMetrics
+)
+
+var (
+	validExtendedMetricsTypes = []ExtendedMetricsType{
+		NoExtendedMetrics,
+		SimpleExtendedMetrics,
+		DetailedExtendedMetrics,
+	}
+)
+
+func (t ExtendedMetricsType) String() string {
+	switch t {
+	case NoExtendedMetrics:
+		return "none"
+	case SimpleExtendedMetrics:
+		return "simple"
+	case DetailedExtendedMetrics:
+		return "detailed"
+	}
+	return "unknown"
+}
+
+// UnmarshalYAML unmarshals an ExtendedMetricsType into a valid type from string.
+func (t *ExtendedMetricsType) UnmarshalYAML(unmarshal func(interface{}) error) error {
```

Comment:
nit: maybe just add a comment specifying that we are implementing the `Unmarshaler` interface for the [yaml.v2](https://godoc.org/gopkg.in/yaml.v2) package.